setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php

```diff
@@ -43,7 +88,14 @@ public function addExcludePatterns(array $excludePatterns)
      */
     public function getList($path)
     {
+
         $realPath = realpath($path);
+        $isGeneration = $this->isGeneration($realPath);
+
+        // Generation folders should not have their results cached since they may actually change during compile
+        if (!$isGeneration && isset($this->fileResults[$realPath])) {
+            return $this->fileResults[$realPath];
+        }
         if (!(bool)$realPath) {
```

Comment:
Oh, well, you'd have to ask the original author of the class.  I didn't write that line.